File: dotnet/SketchOxide/src/MergeableSketch.cs
namespace SketchOxide;

/// <summary>
/// Interface for sketch data structures that support merging.
/// Allows combining multiple sketches into a single sketch for distributed aggregation.
/// </summary>
/// <typeparam name="T">The sketch type that can be merged.</typeparam>
public interface IMergeableSketch<in T> where T : class
{
    /// <summary>
    /// Merges another sketch of the same type into this sketch.
    /// </summary>
    /// <param name="other">The sketch to merge. Must be the same type and compatible configuration.</param>
    /// <exception cref="ArgumentNullException">Thrown if other is null.</exception>
    /// <exception cref="ArgumentException">Thrown if sketches are incompatible (e.g., different precision).</exception>
    /// <exception cref="ObjectDisposedException">Thrown if either sketch has been disposed.</exception>
    void Merge(T other);
}

File: java/src/main/java/com/sketches_oxide/MergeableSketch.java
package com.sketches_oxide;

/**
 * Interface for sketches that support merging.
 *
 * Defines the merge contract for combining multiple sketch instances.
 * This is useful for distributed computing where sketches are computed
 * on different nodes and need to be aggregated.
 *
 * @param <T> the concrete sketch type
 */
public interface MergeableSketch<T> {
    /**
     * Merge another sketch into this one.
     *
     * After merging, this sketch represents the combined state of both sketches.
     * The other sketch is not modified.
     *
     * @param other the sketch to merge
     * @throws IllegalStateException if either sketch is closed
     * @throws IllegalArgumentException if the sketches are incompatible
     *                                  (e.g., different precision levels)
     * @throws NullPointerException if other is null
     */
    void merge(T other);
}
